// @hds/edge-sdk/models/serviceToService/edgeServiceToService.js
"use strict";
/**
 * @module ServiceToService
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdgeServicePublishMessageMode = exports.EdgeServiceMessageType = void 0;
/**
 * Parsing modes supported by the Service to Service Subscription data handler
 */
var EdgeServiceMessageType;
(function (EdgeServiceMessageType) {
    EdgeServiceMessageType["UNHANDLED"] = "UNHANDLED";
    EdgeServiceMessageType["STRING"] = "STRING";
    EdgeServiceMessageType["JSON_OBJECT"] = "JSON_OBJECT";
    EdgeServiceMessageType["EDGE_DATA_OBJECT"] = "EDGE_DATA_OBJECT";
})(EdgeServiceMessageType = exports.EdgeServiceMessageType || (exports.EdgeServiceMessageType = {}));
/**
 * Publish modes supported by the Edge Service to Service Publisher
 */
var EdgeServicePublishMessageMode;
(function (EdgeServicePublishMessageMode) {
    EdgeServicePublishMessageMode["STRING_OR_BUFFER"] = "NONE";
    EdgeServicePublishMessageMode["CONVERT_TO_JSON"] = "CONVERT_TO_JSON";
})(EdgeServicePublishMessageMode = exports.EdgeServicePublishMessageMode || (exports.EdgeServicePublishMessageMode = {}));
