// @hds/edge-sdk/models/api/resultStatusObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResultStatusObject = exports.ResultStatusLegacy = exports.ResultStatus = void 0;
var ResultStatus;
(function (ResultStatus) {
    ResultStatus["OK"] = "OK";
    ResultStatus["WARNING"] = "WARNING";
    ResultStatus["ERROR"] = "ERROR";
})(ResultStatus = exports.ResultStatus || (exports.ResultStatus = {}));
var ResultStatusLegacy;
(function (ResultStatusLegacy) {
    ResultStatusLegacy[ResultStatusLegacy["OK"] = 0] = "OK";
    ResultStatusLegacy[ResultStatusLegacy["WARNING"] = 1] = "WARNING";
    ResultStatusLegacy[ResultStatusLegacy["ERROR"] = 2] = "ERROR";
})(ResultStatusLegacy = exports.ResultStatusLegacy || (exports.ResultStatusLegacy = {}));
class ResultStatusObject {
    constructor(result, status, message) {
        this.status = ResultStatus.OK;
        this.result = result;
        this.status = status || ResultStatus.OK;
        this.message = message;
    }
}
exports.ResultStatusObject = ResultStatusObject;
